build_modules/README.md

```diff
@@ -32,3 +32,70 @@ building your Dart code.
 
 [`build_web_compilers`]: https://pub.dartlang.org/packages/build_web_compilers
 [`build_vm_compilers`]: https://pub.dartlang.org/packages/build_vm_compilers
+
+## Module algorithms
+
+The overall process for emitting modules follows these steps:
+
+1. Emit a `.module.library` asset for every `.dart` library containing a
+   description of the directives (imports, exports, and parts), a list of the
+   `dart:` imports used, and a determination of whether the library is an
+   "entrypoint". Entrypoint libraries are either those which are likely to be
+   directly imported (they are under `lib/` but not `lib/src/`) or which have a
+   `main`. Only the libraries that can be imported (they aren't part files) will
+   have an output. This step is mainly separated out for better invalidation
+   behavior since the output will change less frequently than the code. The
+   outputs from this step are non differentiated by platform.
+1. Emit package level `.meta_module` assets. This is an aggregation of the
+   `.module.library` information with a first run at creating a module
+   structure. The output depends on the algorithm, described below. The outputs
+   from this step are specific to a platform and will exclude any libraries
+   which contain unsupported `dart:` imports for a platform. Conditional imports
+   will also be resolved to a concrete dependency based on the platform. In this
+   step the dependencies of modules references the imported library, which may
+   not match the primary source of the module containing that library.
+1. Emit package level `.meta_module.clean` assets. This step performs an extra
+   run of the strongly connected components algorithm so that any libraries
+   which are in an import cycle are grouped into a single module. This is done
+   as a separate step so that it can be sure it may read all the `.meta_module`
+   results across packages in a dependency cycle so that it may resolve import
+   cycles across packages. As modules are merged due to import cycles the new
+   module becomes the union of their sources and dependencies. Dependencies are
+   resolved to the "primary source" of the module containing the imported
+   library.
+1. Emit library level `.module` assets which contain a filtered view of the
```

Comment:
The `library level` comment here is a bit confusing, I am not sure what that means in this context? Maybe it can just be removed.

Reply:
It was meant to be distinct from "package level" above. I think it does read clearly with this part removed so I dropped it.